### PR TITLE
Add DeepAK8 tagger to NanoAOD

### PR DESCRIPTION
**Pull request by hqucms**
This PR adds the DeepAK8 tagger to NanoAOD. As the DeepAK8 taggers are not stored in MiniAOD until in 103X, they are produced from MiniAOD by using `updateJetCollection`.

### CODE REVIEW

**Comment by arizzi** [comment]
@peruzzim I guess we need some rebasing?

**Comment by arizzi** [comment]
@peruzzim I guess we need some rebasing?

**Comment by arizzi** [comment]
@peruzzim I guess we need some rebasing?

**Comment by clelange** [comment]
Hello,

I haven't seen any discussion on this PR, but strongly support it.

Looking at https://github.com/cms-nanoAOD/cmssw/pull/212/files#diff-8255cc2b641b6438258bddebcd7425ef I see that only a subset of the binarised variables is added:

- `pfDeepBoostedDiscriminatorsJetTags:TvsQCD`
- `pfDeepBoostedDiscriminatorsJetTags:WvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:TvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:WvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:ZHbbvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:ZHccvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:bbvsLight`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:ccvsLight`

I would suggest to also add:

- `pfDeepBoostedDiscriminatorsJetTags:ZvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:ZvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:ZbbvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:HbbvsQCD`
- `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:H4qvsQCD`

This would yield a complete set of variables for use in analysis (e.g. cover the diboson resonances use-case).

If there are serious space/size concerns, I'd suggest to remove the **non**-decorrelated variables. I'm also not sure that the precision stored needs to be 10, but that has to be tested.

**Comment by clelange** [comment]
Documentation for the different variables: https://gitlab.cern.ch/DeepAK8/NNKit/tree/for94X#usage-in-94x-experimental

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 1]
```diff
@@ -178,9 +180,33 @@ def nanoAOD_addDeepFlavourTagFor94X2016(process):
...
+    process.tightJetIdAK8.src = "selectedUpdatedPatJetsAK8Puppi"
+    process.tightJetIdLepVetoAK8.src = "selectedUpdatedPatJetsAK8Puppi"
+    process.slimmedJetsAK8WithUserData.src = "selectedUpdatedPatJetsAK8Puppi"
+    return process
+
+
 def nanoAOD_customizeCommon(process):
     run2_miniAOD_80XLegacy.toModify(process, nanoAOD_addDeepBTagFor80X)
     run2_nanoAOD_94X2016.toModify(process, nanoAOD_addDeepFlavourTagFor94X2016)
+    for modifier in run2_nanoAOD_94X2016, run2_nanoAOD_94XMiniAODv1, run2_nanoAOD_94XMiniAODv2:
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/python/nano_cff.py:207]: This leads to updateJetCollection being called more than once in the same job from other customize functions, leading to possible inconsistencies.
We need to change the model of these customize functions.

Comment (peruzzim) [reply PhysicsTools/NanoAOD/python/nano_cff.py:207]: Correcting myself, for the moment, this updateJetCollection is called on a different collection and with postfix, so for the moment it's ok.

**Comment by peruzzim** [comment]
@hqucms could you please add the 5 discriminators listed by @clelange above?

**Comment by hqucms** [comment]
@clelange 

> * `pfDeepBoostedDiscriminatorsJetTags:ZvsQCD`
> * `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:ZvsQCD`
> * `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:ZbbvsQCD`
> * `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:HbbvsQCD`
> * `pfMassDecorrelatedDeepBoostedDiscriminatorsJetTags:H4qvsQCD`

For the mass-decorrelated tagger, I don't think we need to have Zbb and Hbb separately, right?

**Comment by clelange** [comment]
Hi @hqucms - I talked about that with Loukas (I don't know his github handle) and I understood from the discussion that the output is quite different. Is this not the case?
In an analysis, you would still want to be able to discriminate between H->bb and Z->bb independently of mass, since H and Z have different spins and couple to different particles/resonances.

**Comment by gouskos** [comment]
Hi @clelange I thought that we were talking about W->qq vs Z/H->bb, or Z->qq vs Z->bb. Sorry if I misunderstood.

**Comment by peruzzim** [comment]
Needs to be also activated for other 94X and 102X datasets.
This will be done while re-checking the treatment of AK8 jets in all releases #229.

**Pull request merged**